File: source/ag_tools/nmap_scan.py
import nmap
import sys
from typing import Dict, Any
from langchain_core.tools import tool

def info():
    print("""
Простой сетевой сканер на основе Nmap (учебный пример).
Сканирует указанный хост на наиболее распространённые порты,
выводит их состояние и версии сервисов.
""")


def scan_host(target_host: str) -> Dict[str, Any]:
    # , ports: str = "22,80,443,8080,8443"
    """
    Выполняет сканирование хоста с помощью Nmap.

    :param target_host: IP-адрес или доменное имя цели.
    :param ports: строка с портами для сканирования (формат Nmap: "22,80,443" или "1-1024").
    :return: словарь с результатами сканирования (структура python-nmap).
    """
    # Создаём объект сканера
    scanner = nmap.PortScanner()

    # Аргументы Nmap:
    # -sV — определение версий служб
    # -T4 — агрессивный тайминг (ускоряет сканирование)
    # -Pn — считать хост доступным (пропустить ping-проверку)
    # arguments = f"-sV -T4 -Pn -p {ports}"
    # arguments = '-A -p- -vv --reason --script "default,safe,vuln,discovery" -T4'
    arguments = '-F -n -T5 -A'

    # print(f"[*] Запуск сканирования: {target_host} с параметрами: {arguments}")
    print(f"[*] Запуск сканирования узла ({target_host})...")
    scanner.scan(hosts=target_host, arguments=arguments)

    return scanner

   
# @tool
@tool(description="Run nmap port scan on a target IP with specified ports.")
def run_scan(target: str = "127.0.0.1") -> str:
    #, ports_to_scan: str = "1-65535"
    try:
        scan_res = ""
        scanner = scan_host(target)

        if target not in scanner.all_hosts():
            s = f"[!] Хост {target} не найден в результатах сканирования. Возможно, он недоступен или отфильтрован."
            scan_res += s
            # print(s)
            return

        host_data = scanner[target]
        s = f"Состояние: {host_data.state().upper()}"
        # print(s)
        scan_res += s

        # Перебираем протоколы (tcp/udp)
        for proto in host_data.all_protocols():
            ports = host_data[proto].keys()
            if not ports:
                s = f"Нет открытых {proto} портов"
                # print(s)
                scan_res += s
                continue

            for port in sorted(ports):
                port_info = host_data[proto][port]
                state = port_info.get('state', 'unknown')
                service = port_info.get('name', 'unknown')
                product = port_info.get('product', '')
                version = port_info.get('version', '')
                extrainfo = port_info.get('extrainfo', '')

                # Собираем полную информацию о сервисе
                service_full = service
                if product:
                    service_full += f" ({product}"
                    if version:
                        service_full += f" {version}"
                    if extrainfo:
                        service_full += f" {extrainfo}"
                    service_full += ")"

                scan_res += f"Порт {port}/{proto} {state.upper()} : {service_full}\n"
                # print(f"Порт {port}/{proto} {state.upper()} : {service_full}")

        return scan_res
    except nmap.PortScannerError as e:
        s = f"[!] Ошибка Nmap: {e}"
        # print(s)
        # print("Проверьте, установлен ли Nmap и права доступа (возможно, нужен sudo).")
        return s
    except Exception as e:
        s = f"[!] Непредвиденная ошибка: {e}"
        # print(s)
        return s

# тест флагов nmap для локального запуска без агента
def run(target: str = "127.0.0.1") -> str:
    try:
        scan_res = ""
        scanner = scan_host(target)

        if target not in scanner.all_hosts():
            s = f"[!] Хост {target} не найден в результатах сканирования. Возможно, он недоступен или отфильтрован."
            scan_res += s
            print(s)
            return

        host_data = scanner[target]
        s = f"Состояние: {host_data.state().upper()}"
        print(s)
        scan_res += s

        # Перебираем протоколы (tcp/udp)
        for proto in host_data.all_protocols():
            ports = host_data[proto].keys()
            if not ports:
                s = f"Нет открытых {proto} портов"
                print(s)
                scan_res += s
                continue

            for port in sorted(ports):
                port_info = host_data[proto][port]
                state = port_info.get('state', 'unknown')
                service = port_info.get('name', 'unknown')
                product = port_info.get('product', '')
                version = port_info.get('version', '')
                extrainfo = port_info.get('extrainfo', '')

                # Собираем полную информацию о сервисе
                service_full = service
                if product:
                    service_full += f" ({product}"
                    if version:
                        service_full += f" {version}"
                    if extrainfo:
                        service_full += f" {extrainfo}"
                    service_full += ")"

                scan_res += f"Порт {port}/{proto} {state.upper()} : {service_full}\n"
                print(f"Порт {port}/{proto} {state.upper()} : {service_full}")

        return scan_res
    except nmap.PortScannerError as e:
        s = f"[!] Ошибка Nmap: {e}"
        print(s)
        print("Проверьте, установлен ли Nmap и права доступа (возможно, нужен sudo).")
        return s
    except Exception as e:
        s = f"[!] Непредвиденная ошибка: {e}"
        print(s)
        return s

File: src/agent_tools/nmap_scan.rs
//! Простой сетевой сканер на основе Nmap (учебный пример), как и в Python-версии,
//! НЕ подключён в src/main.rs как активный инструмент агента.
#![allow(dead_code)]

use std::process::Command;

use anyhow::{bail, Context, Result};
use async_trait::async_trait;
use roxmltree::Document;
use serde_json::{json, Value};

use crate::tool::AgentTool;

pub fn info() {
    println!(
        "\nПростой сетевой сканер на основе Nmap (учебный пример).\nСканирует указанный хост на наиболее распространённые порты,\nвыводит их состояние и версии сервисов.\n"
    );
}

struct PortInfo {
    port: u32,
    protocol: String,
    state: String,
    service: String,
}

struct HostScanResult {
    state: String,
    ports: Vec<PortInfo>,
}

/// Выполняет сканирование хоста с помощью Nmap.
fn scan_host(target_host: &str) -> Result<Option<HostScanResult>> {
    println!("[*] Запуск сканирования узла ({target_host})...");

    // Аргументы Nmap: -F (частые порты) -n (без DNS) -T5 (агрессивный тайминг) -A (определение ОС/версий)
    let output = Command::new("nmap")
        .args(["-F", "-n", "-T5", "-A", "-oX", "-", target_host])
        .output()
        .context("не удалось запустить nmap (проверьте, что он установлен)")?;

    if !output.status.success() {
        bail!("nmap завершился с ошибкой: {}", String::from_utf8_lossy(&output.stderr));
    }

    let xml = String::from_utf8_lossy(&output.stdout);
    let doc = Document::parse(&xml).context("не удалось разобрать XML-вывод nmap")?;

    let Some(host_node) = doc.descendants().find(|n| n.has_tag_name("host")) else {
        return Ok(None);
    };

    let state = host_node
        .descendants()
        .find(|n| n.has_tag_name("status"))
        .and_then(|n| n.attribute("state"))
        .unwrap_or("unknown")
        .to_string();

    let mut ports = Vec::new();
    for port_node in host_node.descendants().filter(|n| n.has_tag_name("port")) {
        let protocol = port_node.attribute("protocol").unwrap_or("tcp").to_string();
        let port: u32 = port_node
            .attribute("portid")
            .and_then(|v| v.parse().ok())
            .unwrap_or(0);
        let port_state = port_node
            .children()
            .find(|n| n.has_tag_name("state"))
            .and_then(|n| n.attribute("state"))
            .unwrap_or("unknown")
            .to_string();

        let service_node = port_node.children().find(|n| n.has_tag_name("service"));
        let service_name = service_node.and_then(|n| n.attribute("name")).unwrap_or("unknown");
        let product = service_node.and_then(|n| n.attribute("product")).unwrap_or("");
        let version = service_node.and_then(|n| n.attribute("version")).unwrap_or("");
        let extrainfo = service_node.and_then(|n| n.attribute("extrainfo")).unwrap_or("");

        // Собираем полную информацию о сервисе
        let mut service_full = service_name.to_string();
        if !product.is_empty() {
            service_full += &format!(" ({product}");
            if !version.is_empty() {
                service_full += &format!(" {version}");
            }
            if !extrainfo.is_empty() {
                service_full += &format!(" {extrainfo}");
            }
            service_full += ")";
        }

        ports.push(PortInfo {
            port,
            protocol,
            state: port_state,
            service: service_full,
        });
    }

    Ok(Some(HostScanResult { state, ports }))
}

fn format_scan_result(result: &HostScanResult) -> String {
    let mut out = format!("Состояние: {}", result.state.to_uppercase());
    if result.ports.is_empty() {
        out += "\nНет открытых портов";
    }
    for p in &result.ports {
        out += &format!(
            "\nПорт {}/{} {} : {}",
            p.port,
            p.protocol,
            p.state.to_uppercase(),
            p.service
        );
    }
    out
}

fn run_scan_blocking(target: &str) -> String {
    match scan_host(target) {
        Ok(None) => format!(
            "[!] Хост {target} не найден в результатах сканирования. Возможно, он недоступен или отфильтрован."
        ),
        Ok(Some(result)) => format_scan_result(&result),
        Err(e) => format!("[!] Ошибка Nmap: {e}"),
    }
}

/// Инструмент агента: запуск nmap-сканирования порта на целевом IP.
/// Как и в Python-версии — реализован, но не зарегистрирован в активном списке tools.
pub struct RunScan;

#[async_trait]
impl AgentTool for RunScan {
    fn name(&self) -> &str {
        "run_scan"
    }

    fn schema(&self) -> Value {
        json!({
            "type": "function",
            "function": {
                "name": "run_scan",
                "description": "Run nmap port scan on a target IP with specified ports.",
                "parameters": {
                    "type": "object",
                    "properties": {
                        "target": {"type": "string", "description": "IP-адрес или доменное имя цели"}
                    }
                }
            }
        })
    }

    async fn call(&self, args: Value) -> String {
        let target = args
            .get("target")
            .and_then(Value::as_str)
            .unwrap_or("127.0.0.1")
            .to_string();

        tokio::task::spawn_blocking(move || run_scan_blocking(&target))
            .await
            .unwrap_or_else(|e| format!("[!] Непредвиденная ошибка: {e}"))
    }
}

/// Тест флагов nmap для локального запуска без агента (как `run()` в Python-версии).
pub fn run(target: &str) -> String {
    match scan_host(target) {
        Ok(None) => {
            let s = format!(
                "[!] Хост {target} не найден в результатах сканирования. Возможно, он недоступен или отфильтрован."
            );
            println!("{s}");
            s
        }
        Ok(Some(result)) => {
            let s = format_scan_result(&result);
            println!("{s}");
            s
        }
        Err(e) => {
            let s = format!("[!] Ошибка Nmap: {e}");
            println!("{s}");
            s
        }
    }
}
